### PR TITLE
feat(harvester): add Memovs audiovisual archives API

### PR DESCRIPTION
**Pull request by rerowep**
Add new API harvester for Médiathèque Valais audiovisual archives (archives.memovs.ch). This harvester enables automatic ingestion of Film, Photo, and Audio metadata from the Memovs digital archives.

Implementation includes:
- ApiMemovs class with pagination and date filtering support
- JSON transformation layer mapping Memovs metadata to RERO-ILS
- Document type detection (Film, Audio, Photo, Other)
- Holdings creation with electronic locations
- Support for creators, contributors, subjects, and descriptions
- Integration with existing CLI and task infrastructure

Configuration added as VS-MEMO in apisources.yml with endpoint https://archives.memovs.ch/docs/api/ and code 'memovs'.

### CODE REVIEW

File: rero_ils/modules/api_harvester/cli.py [hunk 1]
```diff
@@ -116,7 +117,8 @@ def info():
         click.echo(f"\tlastrun   : {api.lastrun}")
         click.echo(f"\turl       : {api.url}")
         click.echo(f"\tclassname : {api.classname}")
-        click.echo(f"\tcode   : {api.code}")
+        click.echo(f"\tcode      : {api.code}")
+        click.echo(f"\tsettings  : {api.settings}")
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/api_harvester/cli.py:155]: No we need it

File: rero_ils/modules/api_harvester/utils.py [hunk 2]
```diff
@@ -74,6 +75,9 @@ def api_source(name, url="", classname=None, code="", update=False):
         if code != "":
             source.code = code
             msgs.append(f"code:{code}")
+        if settings is not None:
+            source.settings = settings
+            msgs.append(f"settings:{settings}")
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/api_harvester/utils.py:66]: nop

File: rero_ils/modules/api_harvester/memovs/api.py [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+        settings = self.config.settings or {}
+        location_pid = settings.get("location_pid")
+        item_type_pid = settings.get("item_type_pid")
+        if not (location_pid and item_type_pid):
+            return
+
+        cached = self._items_cache.pop(document_pid, None)
+        if cached is not None:
+            for item_pid in cached.values():
+                if item := Item.get_record_by_pid(item_pid):
```
Comment (jma) [inline rero_ils/modules/api_harvester/memovs/api.py]: Same code as 175, can probably be factorized.

File: rero_ils/modules/api_harvester/utils.py [hunk 4]
```diff
@@ -74,6 +75,9 @@ def api_source(name, url="", classname=None, code="", update=False):
...
         # TODO: commit not working get stuck
```
Comment (jma) [inline rero_ils/modules/api_harvester/utils.py]: To remove?

File: tests/api_harvester/memovs/test_memovs_api.py [hunk 5]
```diff
@@ -0,0 +1,813 @@
...
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Test ApiMemovs helper methods."""
+
+from unittest.mock import MagicMock, patch
+
+from rero_ils.modules.api_harvester.memovs.api import ApiMemovs
+from rero_ils.modules.api_harvester.models import HarvestActionType
+
+_LOCATION_PATCH = "rero_ils.modules.api_harvester.memovs.api.Location.get_record_by_pid"
```
Comment (jma) [inline tests/api_harvester/memovs/test_memovs_api.py:11]: Is this a good approach?

File: tests/api_harvester/memovs/test_memovs_api.py [hunk 6]
```diff
@@ -0,0 +1,813 @@
...
+_MEMOVS_JSON_PATCH = "rero_ils.modules.api_harvester.memovs.api.memovs_json"
+_REQUESTS_PATCH = "rero_ils.modules.api_harvester.memovs.api.requests_retry_session"
+_ILS_INDEXER_PATCH = "rero_ils.modules.api_harvester.memovs.api.IlsRecordsIndexer"
+_HOLDING_PATCH = "rero_ils.modules.api_harvester.memovs.api.Holding"
+_EXTRACTED_REF_PATCH = "rero_ils.modules.api_harvester.memovs.api.extracted_data_from_ref"
+
+
+def _make_harvester(settings):
+    """Return an ApiMemovs instance with a mocked config, bypassing __init__."""
+    harvester = ApiMemovs.__new__(ApiMemovs)
```
Comment (jma) [inline tests/api_harvester/memovs/test_memovs_api.py:29]: Why __new__?

**Review by jma — APPROVED** [review]

File: rero_ils/alembic/4b2a1c3d8e9f_apiharvester_location_itemtype_pids.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2024 RERO
```
Comment (PascalRepond) [inline rero_ils/alembic/4b2a1c3d8e9f_apiharvester_location_itemtype_pids.py]: 2026?

File: rero_ils/alembic/3fa8b1c2d0e7_fix_records_metadata_version_upgrade.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2019 RERO
```
Comment (PascalRepond) [inline rero_ils/alembic/3fa8b1c2d0e7_fix_records_metadata_version_upgrade.py]: 2026?

File: rero_ils/modules/api_harvester/memovs/dojson/json/__init__.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2024 RERO
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/__init__.py]: 2026?

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 10]
```diff
@@ -0,0 +1,802 @@
...
+from requests.exceptions import RequestException
+
+from rero_ils.modules.utils import get_mef_url, get_schema_for_resource, requests_retry_session
+
+_IDREF_URL_RE = re.compile(r"idref\.fr/(\w+)")
+
+# Maps RDA content type code (last segment of @id) to RERO-ILS document type
+_CONTENT_TYPE_MAP = {
+    # Moving image
+    "rdaco:1017": {"main_type": "docmaintype_movie_series", "subtype": "docsubtype_movie"},  # 2D moving image
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: "rdaco:1017" is "tactile notated movement". I would map this as "main_type": "other".

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 11]
```diff
@@ -0,0 +1,802 @@
...
+    "rdaco:1018": {"main_type": "docmaintype_movie_series", "subtype": "docsubtype_movie"},  # 3D moving image
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: rdaco:1018 is "tactile text", are you sure this is relevant for this source? Moreover, the mapped type should not be movie but rather docmaintype_book / docsubtype_braille_book

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 12]
```diff
@@ -0,0 +1,802 @@
...
+    "rdaco:1023": {"main_type": "docmaintype_movie_series", "subtype": "docsubtype_movie"},  # image animée 2D
+    # Audio
+    "rdaco:1011": {"main_type": "docmaintype_audio", "subtype": "docsubtype_music"},  # musique exécutée
+    "rdaco:1012": {"main_type": "docmaintype_audio", "subtype": "docsubtype_sound"},  # sons
+    "rdaco:1013": {"main_type": "docmaintype_audio", "subtype": "docsubtype_recorded_words"},  # parole énoncée
+    # Still image
+    "rdaco:1014": {"main_type": "docmaintype_image", "subtype": "docsubtype_photography"},  # image fixe
+    "rdaco:1016": {"main_type": "docmaintype_image", "subtype": "docsubtype_photography"},  # still image
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: rdaco:1016 is "tactile notated music", not still image

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 13]
```diff
@@ -0,0 +1,802 @@
...
+            agent = contribution.get("bf:agent", {})
+            if not isinstance(agent, dict):
+                continue
+
+            agent_label = agent.get("rdfs:label", "")
+            if agent_label.startswith(("http://", "https://")):
+                current_app.logger.warning(
+                    "Memovs %s: bf:contribution agent has a URL as rdfs:label (%s).",
+                    self.memovs_id,
+                    agent_label,
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py:336]: This seems very specific verification ?

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 14]
```diff
@@ -0,0 +1,802 @@
...
+                return
+
+        # Last resort: dateless stub so the required field is always present
+        self.json_dict["provisionActivity"] = [
+            {
+                "statement": [
+                    {"label": [{"value": "Médiathèque Valais"}], "type": "bf:Agent"},
+                ],
+                "type": "bf:Production",
+            }
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: I don't think that would work, provisionActivity.startDate is required.

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 15]
```diff
@@ -0,0 +1,802 @@
...
+                        "statement": [
+                            {"label": [{"value": "Médiathèque Valais"}], "type": "bf:Agent"},
+                            {"label": [{"value": str(year)}], "type": "Date"},
+                        ],
+                        "type": "bf:Production",
+                    }
+                ]
+                return
+
+        # Fallback: try dcterms:created or dcterms:modified
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: Is this fallback really needed ?

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 16]
```diff
@@ -0,0 +1,802 @@
...
+    def trans_admin_metadata(self):
+        """Placeholder for future adminMetadata transformations.
+
+        bf:note labels are written to a LocalField record by ApiMemovs.sync_local_field.
+        """
+
+    def trans_notes(self):
+        """Map source fields to document notes.
+
+        - bf:provenance → noteType ``general``
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: What's this? "bf:provenance" is not a valid bibframe class

File: rero_ils/modules/api_harvester/memovs/dojson/__init__.py [hunk 17]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2024 RERO
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/__init__.py]: 2026? Check all files

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_role-v0.0.1.json [hunk 18]
```diff
@@ -135,6 +134,7 @@
       "rdd",
       "res",
       "rpc",
+      "rpt",
```
Comment (PascalRepond) [inline rero_ils/modules/documents/jsonschemas/documents/document_contribution_role-v0.0.1.json:137]: To be checked but I don't think rpt is a valid RDA agent role code! https://access.rdatoolkit.org/Glossary?externalId=en-US_rdaregistry.info-Elements-w-P10437&highlight=role

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_role-v0.0.1.json [hunk 19]
```diff
@@ -87,7 +87,6 @@
       "fmp",
       "his",
       "hst",
-      "ill",
```
Comment (PascalRepond) [inline rero_ils/modules/documents/jsonschemas/documents/document_contribution_role-v0.0.1.json:90]: Why??

**Review by PascalRepond — CHANGES_REQUESTED** [review]

File: tests/api_harvester/memovs/test_memovs_api.py (hunk 5 above)
Comment (rerowep) [reply tests/api_harvester/memovs/test_memovs_api.py:11]: For this specific file, patches are the right call. Here's why:

The harvester's value is its orchestration logic — deciding when to create, update, or delete, managing caches, handling edge cases. That logic is what these tests cover, and mocks let you test it in isolation without needing a running DB, ES, and Flask app context.

The CLAUDE.md guideline backs this up: "only test specific app behaviour and not the behaviour of external modules (e.g. invenio dependencies)." Item.create, Document.get_record_by_pid, etc. are Invenio — you'd be testing Invenio, not the harvester.

File: tests/api_harvester/memovs/test_memovs_api.py (hunk 6 above)
Comment (rerowep) [reply tests/api_harvester/memovs/test_memovs_api.py:29]: Because ApiMemovs.__init__ likely does real work — connecting to the DB, loading config from SQLAlchemy, etc. — that would require a full app/DB setup to run. __new__ skips __init__ entirely, giving you a bare instance you can then configure manually with just the fields the tests need.

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py (hunk 11 above)
Comment (rerowep) [reply rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: rdaco:1011 → audio / music        (2,128 records)
rdaco:1012 → audio / sound        (9 records)
rdaco:1013 → audio / recorded_words (4,341 records)
rdaco:1014 → image / photography  (34,375 records)
rdaco:1023 → movie_series / movie (29,670 records)

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py (hunk 13 above)
Comment (rerowep) [reply rero_ils/modules/api_harvester/memovs/dojson/json/model.py:336]: Normal rdfs:label values are person/organisation names in "Lastname, Firstname" format or freeform strings like "Chef de l'expédition : Raymond Lambert...".

Why the URL check exists: 8 records in the live data have corrupted entries where rdfs:label contains GIF image URLs — apparent data entry errors in the legacy Memovs system (msav.memovs.ch). Examples:


avnid:11272  http://msav.memovs.ch/images/retour.gifhttp://msav.memovs.ch/images/save.gif
avnid:549    http://msav.memovs.ch/images/save.gif
avnid:29263  http://msav.memovs.ch/images/button_tel.gifhttp://msav.memovs.ch/images/button_tel.gif
These are old web interface button images that got accidentally stored as the agent name — probably a copy/paste or migration artefact from the old msav.memovs.ch system. The check skips them and logs a warning rather than creating a contribution with a URL as the person's name.

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py (hunk 14 above)
Comment (rerowep) [reply rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: startDate is only required when type === 'bf:Publication', enforced by a formly UI expression, not by the JSON schema required array
The code always uses type: bf:Production, so startDate is never required
The dateless stub at line 404 (the one you highlighted) is the last resort fallback reached only when Memovs provides no date at all in bf:provisionActivity, dcterms:created, or dcterms:modified
The two paths above it (lines 375 and 391) always include startDate when a year is found

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_role-v0.0.1.json (hunk 19 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document_contribution_role-v0.0.1.json:90]: It is twice in the list.

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py (hunk 15 above)
Comment (rerowep) [reply rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: 70,674 records — year found in bf:provisionActivity ✓
194 records — bf:provisionActivity exists but has empty rdf:value and rdfs:label, so the fallback to dcterms:created rescues the year
0 records — reach the dateless stub (so that last-resort path is never hit in practice)
The fallback pattern for those 194 records is always the same:


"bf:provisionActivity": {"bf:date": {"rdf:value": "", "rdfs:label": ""}},
"dcterms:created": "2007-04-05T07:46:52.000Z"
Empty date in the provision activity, but the record creation timestamp has a usable year. The fallback is correct and necessary.

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_role-v0.0.1.json (hunk 18 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document_contribution_role-v0.0.1.json:137]: rpt is not an RDA role — but the data is using MARC relators intentionally via BIBFRAME, and rpt is a valid entry in that vocabulary. Memovs uses bf:role with LoC relator URIs (e.g. http://id.loc.gov/vocabulary/relators/rpt), not RDA role URIs. That's standard BIBFRAME practice. All the other codes in the Memovs data (pht, prf, ctb, pro, drt…) are equally MARC relators, not RDA roles.

**Review by PascalRepond — APPROVED** [review]

**Review by jma — APPROVED** [review]

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 20]
```diff
@@ -0,0 +1,827 @@
...
+        if not year and (date_str := (self.data.get("dcterms:created", "") or self.data.get("dcterms:modified", ""))):
+            if match := re.search(r"(\d{4})", str(date_str)):
+                year = int(match[1])
+
+        statement = []
+        if agent := bf_provision.get("bf:agent", {}):
+            if agent_label := agent.get("rdfs:label", "") if isinstance(agent, dict) else "":
+                statement.append({"label": [{"value": agent_label}], "type": "bf:Agent"})
+
+        provision_activity = {"type": "bf:Production"}
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: use: bf_provision.get("@type")

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 21]
```diff
@@ -0,0 +1,827 @@
...
+        the source actually exports one — no institution name is fabricated.
+        provisionActivity is required by the schema and is always set.
+        """
+        bf_provision = self.data.get("bf:provisionActivity", {})
+        if not isinstance(bf_provision, dict):
+            bf_provision = {}
+
+        year = None
+        date_label = ""
+        if bf_provision.get("@type") == "bf:Publication":
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: if probably not needed

File: rero_ils/modules/api_harvester/memovs/api.py [hunk 22]
```diff
@@ -0,0 +1,530 @@
...
+
+        When ``_lofis_cache`` is populated (batch harvest mode) the ES existence
+        query is replaced by a dict lookup and writes are deferred to the next
+        ``_flush_pending`` call.
+
+        :param document_pid: the RERO-ILS document PID.
+        :param notes: list of label strings extracted from bf:note.
+        """
+        settings = self.config.settings or {}
+        field_name = settings.get("local_field_name", "field_2")
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/api.py:245]: Why everything in field2 ? each element should be in its own field

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 23]
```diff
@@ -0,0 +1,756 @@
...
+        # a structured place (country) is only relevant for a publication
+        if place_label and provision_type == "bf:Publication":
+            # the source never exports a country code; default to "xx" (undetermined)
+            provision_activity["place"] = [{"country": "xx"}]
+        if year:
+            statement.append({"label": [{"value": date_label or str(year)}], "type": "Date"})
+            provision_activity["startDate"] = year
+        else:
+            provision_activity["startDate"] = 9999
+            provision_activity["note"] = "Date(s) uncertain or unknown"
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py:431]: "Date(s) incertaine(s) ou inconnue(s)" in French

**Comment by PascalRepond** [comment]
## Latest user tests

- [x] field `content`, do not take the timestamps into account [image](https://github.com/user-attachments/assets/32dabc84-0369-4bb8-8d20-9c1db9850d4d)
- [x] Take into account backspaces. Example: https://archives.memovs.ch/docs/id/s068-007-032b.json, `\r\n` should become a backspace in the ILS field
- [x] Location of the items should be pid:1223
- [x] `bf:partOf` if `bf:noteType`= 'fonds' OR 'emission' => ignore this partOf item

File: rero_ils/modules/api_harvester/memovs/dojson/json/model.py [hunk 24]
```diff
@@ -0,0 +1,775 @@
...
+        the RDA content type code (e.g. film, photography, audio), falling
+        back to webSite when the code is absent or unrecognised.
+        """
+        electronic_locators = []
+
+        if bf_elocs := self.data.get("bf:electronicLocator", []):
+            for eloc in bf_elocs:
+                url = eloc.get("@id", "")
+                note_type = eloc.get("bf:noteType", "")
+                label = eloc.get("rdfs:label", "")
```
Comment (PascalRepond) [inline rero_ils/modules/api_harvester/memovs/dojson/json/model.py]: We don't need to take the electronicLocator label into account, delete this line and both following uses of the `label` variable.
